### PR TITLE
LibSQL: Data Storage 

### PR DESCRIPTION
**Pull request by JanDeVisser**
This is the beginning of the LibSQL data storage functionality. Data is stored in a Heap object, which is a block based file.
BTrees (and their constituent nodes and keys) and Tuples can be stored in a Heap. A very simple Database object manages
rudimentary SQL tables in a Heap.

With apologies for the monster patch. I'm new to this github stuff and the way this project works, but I promise I will learn...

### CODE REVIEW

File: Userland/Libraries/LibSQL/TreeNode.cpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    dbgln("[{}] {} >= {} ({} -> {})",
+        m_pointer, (String) key, (String) m_entries[size() - 1],
+        size(), m_down_nodes[size()]->pointer());
+    return down_node(size())->node_for(key);
+}
+
+Optional<u32> TreeNode::get(Key& key)
+{
+    dbgln("[{}] find({})",
+        m_pointer, (String) key);
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/TreeNode.cpp]: is there not a better way than to c-cast it?

File: Userland/Libraries/LibSQL/TreeNode.cpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            return down_node(ix)->node_for(key);
+        }
+    }
+    dbgln("[{}] {} >= {} ({} -> {})",
+        m_pointer, (String) key, (String) m_entries[size() - 1],
+        size(), m_down_nodes[size()]->pointer());
+    return down_node(size())->node_for(key);
+}
+
+Optional<u32> TreeNode::get(Key& key)
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/TreeNode.cpp:227]: is key supposed to cahnge in this function?

if yes:
dont use a reference
if no: 
mark it as constant

I dont think a getter should change the value of the provided index/key

File: Userland/Libraries/LibSQL/TreeNode.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+void TreeNode::serialize(ByteBuffer& buffer) const
+{
+    u32 sz = size();
+    serialize_to(buffer, sz);
+    for (auto ix = 0u; ix < size(); ix++) {
+        auto& entry = m_entries[ix];
+        serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[ix]);
+        entry.serialize(buffer);
+    }
+    serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[size()]);
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/TreeNode.cpp]: you can just append a `u` to a number to make it unsigned, like `0u`

File: Userland/Libraries/LibSQL/TreeNode.cpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+void TreeNode::dump() {
+    StringBuilder builder;
+    builder.appendff("[{}] size {}", pointer(), (int) size());
+    if (is_leaf()) {
+        builder.append(" is leaf");
+    } else {
+        builder.appendff(" down nodes {}", m_down_nodes.size());
+    }
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/TreeNode.cpp]: when all paths of the comparison are just one line, we usually omit the curly braces

you might have to fix it on earlyer lines too, this is the first time that I conciously see it

File: Userland/Libraries/LibSQL/Tuple.cpp [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ * Copyright (c) 2021, Jan de Visser <jan@de-visser.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <cstring>
+#include <LibSQL/Tuple.h>
+
+namespace SQL {
+
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/Tuple.cpp:13]: empty?

File: Userland/Libraries/LibSQL/Type.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+#define ENUMERATE_SQL_TYPES \
+    __ENUMERATE_SQL_TYPE("text", Text, String)                   \
+    __ENUMERATE_SQL_TYPE("int", Integer, int)                    \
+    __ENUMERATE_SQL_TYPE("float", Float, double)
+
+enum SQLType {
+#define __ENUMERATE_SQL_TYPE(name, type, impl) type,
+    ENUMERATE_SQL_TYPES
+#undef __ENUMERATE_SQL_TYPE
+};
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/Type.h]: this looks weird

File: Userland/Libraries/LibSQL/Meta.cpp [hunk 7]
```diff
@@ -10,78 +10,112 @@
 
 namespace SQL {
 
-ColumnDef::ColumnDef(String const& name, SQLType sql_type)
-    : Relation(name)
+/* ----------------------------------------------------------------------- */
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/Meta.cpp:13]: why do you use those lines?

File: Userland/Libraries/LibSQL/TreeNode.cpp [hunk 8]
```diff
@@ -94,15 +97,21 @@ bool TreeNode::insert(Key const& key, TreeNode *right)
         }
     }
 
+#ifdef BTREE_DEBUG
     dbgln("Inserting in this node");
+#endif
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/TreeNode.cpp]: there is something called `dbgln_if`
it essentially does the same, but looks neater and is the preferred way of doing this

File: Tests/LibSQL/TestSQLStorage.cpp [hunk 9]
```diff
@@ -100,6 +101,194 @@ TEST_CASE(create_heap) {
...
+                    break;
+                }
+            }
+            prev = key;
+        }
+        EXPECT_EQ(count, num_keys);
+    }
+}
+
+TEST_CASE(btree_one_key)
```
Comment (Hendiadyoin1) [inline Tests/LibSQL/TestSQLStorage.cpp:339]: why do you make 10 tests doing essentially the same?

File: Tests/LibSQL/TestSQLStorage.cpp [hunk 10]
```diff
@@ -100,6 +101,194 @@ TEST_CASE(create_heap) {
...
+{
+    insert_into_and_scan_btree(15);
+}
+
+TEST_CASE(btree_scan_50_keys)
+{
+    insert_into_and_scan_btree(50);
+}
+
+#if 0
```
Comment (Hendiadyoin1) [inline Tests/LibSQL/TestSQLStorage.cpp]: whats with this unused code

**Review by Hendiadyoin1 — COMMENTED** [review]
This is a very shallow review for now, and I stoped like half way through, but there are some things that should definetly be changed

please cleanup your git history:
there are a lot of commits that essentially say "Test and fix" and you create files, just to delete them 2 commits later
in addition to that the commits arent very atomic, you *test and fix* instead of `fix` then `test to prevent regression` like it is usually done
To the point I have read your history should look something like:
1. LibSQL: Add BTree
2. LibSQL: Add Storage functionality
3. Tests+LibSQL:  Add Tests

lastly please add newlines at the end of every file

**Review by alimpfard — CHANGES_REQUESTED** [review]
This seems like a nice addition, but please observe the contribution guidelines and code style requirements, and clean up the commit history:
https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md

File: Userland/Libraries/LibSQL/BTree.h [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+{
+    ByteBuffer buffer(serialize(leaf.size()));
+    for (auto &node : leaf.nodes()) {
+        buffer += serialize(node);
+    }
+    return buffer;
+}
+
+template <typename KeyType>
+int deserialize(TreeNode<KeyType> &leaf, ByteBuffer const &buffer, int offset)
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h]: Thanks- will look into.

File: Userland/Libraries/LibSQL/Value.h [hunk 12]
```diff
@@ -71,6 +72,7 @@ class Value {
     Function<void(ByteBuffer&, size_t &offset)> m_deserialize;
     Function<size_t()> m_size;
     Function<const char *()> m_type_name;
+    Function<bool(Value const&)> m_can_cast;
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Value.h:75]: Yes. I really wanted to be able to pass Values by reference as stack objects and be able to store them in containers as objects. If I would have used virtual methods this would have been impossible.

File: Userland/Libraries/LibSQL/ReadMe.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
+### SQL Storage Overview
+
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/ReadMe.md:2]: OK

File: Userland/Libraries/LibSQL/Value.cpp [hunk 14]
```diff
@@ -313,6 +313,7 @@ void Value::setup_text()
     };
 
     m_can_cast = [](Value const& other) -> bool {
+        (void) other;
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Value.cpp:316]: OK

File: Userland/Libraries/LibSQL/BTree.cpp [hunk 15]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2021, Jan de Visser <jan@de-visser.net>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "BTree.h"
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.cpp]: Thanks- as I mention below my git/github skills leave to be desired, I'll attempt to clean stuff up so you guys don't have to look at things that have been gone for weeks

File: Userland/Libraries/LibSQL/BTree.h [hunk 16]
```diff
@@ -36,7 +37,7 @@ class BlockFile {
     {
         if (!(*m_file).seek(block * BLOCKSIZE))
         {
-            warnln("Could not seek block %d from file %s", block, m_name);
+//            warnln("Could not seek block %d from file %s", block, m_name);
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h]: I was unaware of this. Will investigate but I don't doubt I can use it.

File: Userland/Libraries/LibSQL/BTree.h [hunk 17]
```diff
@@ -131,162 +148,336 @@ class TreeNodeEntry {
...
 
 private:
-    bool m_is_leaf { true };
-    RefPtr<BTree<KeyType>> m_tree;
+    BTree<KeyType> *m_tree;
     TreeNode<KeyType> *m_up;
     u32 m_pointer;
     Vector<TreeNodeEntry<KeyType>> m_entries;
     Vector<u32> m_down_pointers;
     Vector<TreeNode<KeyType> *> m_down_nodes;
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h]: Later revisions use smart pointers.

File: Userland/Libraries/LibSQL/BTree.h [hunk 18]
```diff
@@ -6,479 +6,124 @@
...
-    String m_name;
-    bool m_duplicates_allowed { false };
-//    OwnPtr<BlockFile> m_blockfile;
-    TreeNode<KeyType> *m_root { nullptr };
-    u32 m_next_record_pointer { 0 };
+    friend BTree;
 };
 
-}
+} // namespace SQL
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h:175]: Will remove. I use it so I don't get confused when I'm missing closing braces.

File: Userland/Libraries/LibSQL/CMakeLists.txt [hunk 19]
```diff
@@ -2,7 +2,16 @@ set(SOURCES
...
+
     BTree.cpp
+    BTreeIterator.cpp
+    Database.cpp
+    Heap.cpp
+    Key.cpp
+    Meta.cpp
+    TreeNode.cpp
+    Tuple.cpp
+    Value.cpp
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/CMakeLists.txt]: Will fix

File: Userland/Libraries/LibSQL/Key.cpp [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+#include <AK/RefPtr.h>
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+
+#include <LibSQL/Key.h>
+#include <LibSQL/Serialize.h>
+
+namespace SQL {
+
+Key::Key()
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Key.cpp:20]: I'm not sure what the guideline is as to what can go in the header and what not. I basically moved everything out-of-line, but if there's stuff that should go inline I'm more than willing to move.

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 1 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: Sure. I will create a to_string() method and have the (String) operator use that.

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 2 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/TreeNode.cpp:227]: Hrm. It doesn't change the value per se, it sets the pointer it finds for that key in the key. So if the key is present in the node, the return value is set in the key as well.

More that willing to change name or semantics, just not sure where to go.

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 3 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: Done

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 4 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: 30 years of habit don't go away that easily, young padawan :-)

I know there's discussions bordering on holy wars on what's best here; I've always used braces but it's not worth the aggravation. I'll remove them.

File: Userland/Libraries/LibSQL/Tuple.cpp (hunk 5 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Tuple.cpp:13]: Not anymore.

File: Userland/Libraries/LibSQL/Type.h (hunk 6 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Type.h]: Isn't this the pattern Andreas uses?

File: Userland/Libraries/LibSQL/Meta.cpp (hunk 7 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Meta.cpp:13]: I use them to find my way around my source. I will delete them

File: Tests/LibSQL/TestSQLStorage.cpp (hunk 9 above)
Comment (JanDeVisser) [reply Tests/LibSQL/TestSQLStorage.cpp:339]: Because it is my experience that with datastructures like this it's impossible to design your test so you find all corner cases. Doing it this way is a bit of a blunt force instrument but I've had good experiences with it, including in this code.

File: Userland/Libraries/LibSQL/Key.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (Hendiadyoin1) [reply Userland/Libraries/LibSQL/Key.cpp:20]: as long as it isn't dependent on functions after it, it usually is inline

when the functions get to long they are moved out of line too

File: Userland/Libraries/LibSQL/TreeNode.cpp [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+        m_pointer, (String) key, ((right) ? right->m_pointer : -1));
+    if (!is_leaf()) {
+        dbgln("Node is not a leaf. Find leaf to insert into");
+        return node_for(key)->insert(key);
+    }
+    VERIFY(is_leaf());
+    if (!m_tree->duplicates_allowed()) {
+        for (auto& entry : m_entries) {
+            if (key == entry) {
+                dbgln("BTree does not allow duplicates and key {} already exists");
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibSQL/TreeNode.cpp]: how did this compile, its missing an argument to fill in the fromat

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 1 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: You don't need to override the operator nor use a c-cast here if you make a to_string() function, just call it directly

File: Userland/Libraries/LibSQL/Type.h (hunk 6 above)
Comment (linusg) [reply Userland/Libraries/LibSQL/Type.h]: yes, it is. not necessarily weird, but haven't looked at it in detail yet.

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 2 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibSQL/TreeNode.cpp:227]: again seems weir for a `get(...)`, but OK

File: Userland/Libraries/LibSQL/BTree.h [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+        TreeNodeEntry<KeyType> node;
+        offset += deserialize(node, buffer, offset);
+        leaf.insert(node);
+    }
+    return offset - old_offset;
+}
+
+template <typename KeyType>
+class BTree {
+public:
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h:129]: Maybe, but in this particular case integration in the actual storage is important. What I mean by that is that the tree lives in persistent storage and only the parts needed for the job at hand should be loaded into memory (it doesn't do a very good jib at that yet but that's beside the point). Also, the keys don't refer to a living, breathing C++ object, but to another piece of data of arbitrary type also living in persistent storage.

So while it may be possible to come up with something that does double duty satisfying these requirements and being a general-purpose in-memory data structure, that's not the problem I'm trying to solve here.

File: Userland/Libraries/LibSQL/BTree.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h]: reading again I now see you probably meant 'tree' instead of 'free', and in that case let me explain. The tree lives in persistent storage. As you traverse it to find keys or iterate over all keys, you need to load nodes from storage, and if you insert keys, update what they point to, or delete keys, you need flush the new tree structure to storage.

This is done using the serialization and deserialization methods.

File: Userland/Libraries/LibSQL/BTree.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibSQL/BTree.h:129]: Sure.

File: Userland/Libraries/LibSQL/BTree.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Libraries/LibSQL/BTree.h]: I actually meant "free" - as in not part of a specific class.

File: Userland/Libraries/LibSQL/BTree.cpp [hunk 23]
```diff
@@ -40,37 +45,56 @@ BTreeIterator BTree::end()
...
 TreeNode * BTree::new_root(u32 pointer) {
     auto old_root = m_root.ptr();
-    m_pointer = (pointer == 0) ? new_record_pointer() : pointer;
-    if (old_root) {
-        m_root = make<TreeNode>(*this, nullptr, old_root, m_pointer);
+    if (pointer) {
+        auto buffer = read_block(pointer);
+        size_t offset = 0;
+        m_root = make<TreeNode>(*this, nullptr, pointer, buffer, offset);
+        m_pointer = pointer;
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.cpp]: Yeah, I pondered what to call it, and in the end I settled on 'pointer' even though it's a horribly overloaded term. What it is is an address in persistent storage the key 'points' to. Other possible names would be things like 'record number', 'block number', etc. None of the alternatives seemed general enough for what I wanted to express so I settled on 'pointer',

File: Userland/Libraries/LibSQL/Key.cpp [hunk 24]
```diff
@@ -40,24 +40,33 @@ Key::Key(IndexDef const& index, ByteBuffer& buffer, size_t& offset)
...
+    dbgln("Serializing key pointer {}", pointer());
+#endif
     serialize_to<u32>(buffer, pointer());
-    for (auto ix = 0u; ix < m_key_data.size(); ix++) {
-        auto &key_part = m_key_data[ix];
-//        auto str_opt = key_part.to_string();
-//        auto &key_part_definition = m_part_definitions[ix];
-//        dbgln("Serializing key[{}] = {}", key_part_definition.name(), (str_opt.has_value()) ? str_opt.value() : "(null)");
+    for (auto &key_part : m_key_data) {
+#ifdef SERIALIZE_DEBUG
```
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Key.cpp:65]: I was pointed towards the dbgln_if functionality,

File: Userland/Libraries/LibSQL/Type.h (hunk 6 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibSQL/Type.h]: similar, but not quite
see 
https://github.com/SerenityOS/serenity/blob/master/Kernel/API/Syscall.h#L31

so it would be
```cpp
#define ENUMERATE_SQL_TYPES(S)   \
     S("text", Text, String)                     \
     S("int", Integer, int)                        \
     S("float", Float, double)

enum SQLType {
#define __ENUMERATE_SQL_TYPE(name, type, impl) type,
    ENUMERATE_SQL_TYPES(__ENUMERATE_SQL_TYPE)
#undef __ENUMERATE_SQL_TYPE
};

```

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 21 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: Not as far as I know, would need to ask on discord myself...

File: Userland/Libraries/LibSQL/BTree.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/BTree.h]: Ah OK. They're mostly gone; I only have template overrides for basic types now.

File: Userland/Libraries/LibSQL/Type.h (hunk 6 above)
Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/Type.h]: Ah OK will fix

File: Userland/Libraries/LibSQL/TreeNode.cpp (hunk 21 above)
Comment (alimpfard) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: Is your host compiler clang?
The format checks aren't enabled on clang

Comment (JanDeVisser) [reply Userland/Libraries/LibSQL/TreeNode.cpp]: gcc 10.2

**Comment by JanDeVisser** [comment]
I think I messed it up even more now. I rebased from the last commit before my first commit, and squashed everything into one commit. But it dragged in all other commits because I synced my tree every now and then.

I'm going to close this and start over.

Big learning experience here. Apologies you have to be part of it.
